### PR TITLE
Adding `Bakeable` trait and derive

### PR DESCRIPTION
**Pull request by robertbastian**

### CODE REVIEW

File: experimental/crabbake/src/bakeable.rs
Comment (Manishearth) [inline experimental/crabbake/src/bakeable.rs:38]: quote supports macro-style list interpolation, so you can write something like `&[#(data),*]` and not have to use the for loop above

File: experimental/crabbake/src/lib.rs
Comment (Manishearth) [inline experimental/crabbake/src/lib.rs:13]: crabbake should not know anything about data providers, this should happen at a separate layer. You should be defining a custom derive  instead

also the trait cannot live in the same crate as the custom derive, use `crabbake/derive`

Comment (Manishearth) [reply experimental/crabbake/src/lib.rs:13]: (we may want the static data provider generation to be something the datagen tool does, too, but having a proc macro for it is decent too)

**Review by Manishearth — COMMENTED** [review]
quick initial comments before you go too far ahead

File: experimental/crabbake/src/lib.rs
Comment (robertbastian) [reply experimental/crabbake/src/lib.rs:13]: I know, I was just playing around with this and didn't want to throw this code away.

How would the datagen work? It would create a .rs file? That seems more brittle than a proc macro.

Comment (Manishearth) [reply experimental/crabbake/src/lib.rs]: Ah, I see.

I think folks might want both, potentially. They can use the same code.

But for now just having const data provider seems ok

**Comment by robertbastian** [comment]
`cargo expand --example const_provider`:

```rust
#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2018::*;
#[macro_use]
extern crate std;
use icu_locid_macros::langid;
use icu_provider::prelude::*;
fn main() {
    {
        struct ConstProvider {
            core_helloworld_1: &'static [(
                icu_provider::prelude::ResourceOptions,
                icu_provider::hello_world::HelloWorldV1<'static>,
            )],
        }
        impl icu_provider::prelude::DataProvider<icu_provider::hello_world::HelloWorldV1Marker>
            for ConstProvider
        {
            fn load_payload(
                &self,
                req: &icu_provider::prelude::DataRequest,
            ) -> Result<
                icu_provider::prelude::DataResponse<icu_provider::hello_world::HelloWorldV1Marker>,
                icu_provider::prelude::DataError,
            > {
                use icu_provider::prelude::*;
                use icu_provider::yoke::ZeroCopyFrom;
                let data: &[(ResourceOptions, icu_provider::hello_world::HelloWorldV1)] = match (
                    req.resource_path.key.category,
                    req.resource_path.key.sub_category.as_unsigned(),
                    req.resource_path.key.version,
                ) {
                    (
                        icu_provider::prelude::ResourceCategory::Core,
                        474237142587688335730024u128,
                        1u16,
                    ) => self.core_helloworld_1,
                    _ => return Err(DataError::MissingResourceKey(req.resource_path.key.clone())),
                };
                #[allow(unreachable_code)]
                match data.binary_search_by_key(&&req.resource_path.options, |(k, _)| k) {
                    Err(_) => return Err(DataError::MissingResourceOptions(req.clone())),
                    Ok(index) => Ok(DataResponse {
                        metadata: DataResponseMetadata::default(),
                        payload: Some(DataPayload::from_owned(
                            <icu_provider::hello_world::HelloWorldV1>::zero_copy_from(
                                &data[index].1,
                            ),
                        )),
                    }),
                }
            }
        }
        const DATA_PROVIDER: ConstProvider = {
            extern crate alloc;
            use icu_provider::hello_world::HelloWorldV1;
            ConstProvider {
                core_helloworld_1: &[
                    (
                        icu_provider::prelude::ResourceOptions {
                            variant: None,
                            langid: Some(icu_locid::LanguageIdentifier {
                                language: unsafe {
                                    icu_locid::subtags::Language::from_raw_unchecked(Some(28258))
                                },
                                script: None,
                                region: None,
                                variants: unsafe {
                                    icu_locid::subtags::Variants::from_raw_unchecked(None)
                                },
                            }),
                        },
                        icu_provider::hello_world::HelloWorldV1 {
                            message: alloc::borrow::Cow::Borrowed("ওহে বিশ\u{9cd}ব"),
                        },
                    ),
                    (
                        icu_provider::prelude::ResourceOptions {
                            variant: None,
                            langid: Some(icu_locid::LanguageIdentifier {
                                language: unsafe {
                                    icu_locid::subtags::Language::from_raw_unchecked(Some(29539))
                                },
                                script: None,
                                region: None,
                                variants: unsafe {
                                    icu_locid::subtags::Variants::from_raw_unchecked(None)
                                },
                            }),
                        },
                        icu_provider::hello_world::HelloWorldV1 {
                            message: alloc::borrow::Cow::Borrowed("Ahoj světe"),
                        },
                    ),
                    (
                        icu_provider::prelude::ResourceOptions {
                            variant: None,
                            langid: Some(icu_locid::LanguageIdentifier {
                                language: unsafe {
                                    icu_locid::subtags::Language::from_raw_unchecked(Some(25956))
                                },
                                script: None,
                                region: None,
                                variants: unsafe {
                                    icu_locid::subtags::Variants::from_raw_unchecked(None)
                                },
                            }),
                        },
                        icu_provider::hello_world::HelloWorldV1 {
                            message: alloc::borrow::Cow::Borrowed("Hallo Welt"),
                        },
                    ),
                    (
                        icu_provider::prelude::ResourceOptions {
                            variant: None,
                            langid: Some(icu_locid::LanguageIdentifier {
                                language: unsafe {
                                    icu_locid::subtags::Language::from_raw_unchecked(Some(27749))
                                },
                                script: None,
                                region: None,
                                variants: unsafe {
                                    icu_locid::subtags::Variants::from_raw_unchecked(None)
                                },
                            }),
                        },
                        icu_provider::hello_world::HelloWorldV1 {
                            message: alloc::borrow::Cow::Borrowed("Καλημέρα κόσμε"),
                        },
                    ),
                    (
                        icu_provider::prelude::ResourceOptions {
                            variant: None,
                            langid: Some(icu_locid::LanguageIdentifier {
                                language: unsafe {
                                    icu_locid::subtags::Language::from_raw_unchecked(Some(28261))
                                },
                                script: None,
                                region: None,
                                variants: unsafe {
                                    icu_locid::subtags::Variants::from_raw_unchecked(None)
                                },
                            }),
                        },
                        icu_provider::hello_world::HelloWorldV1 {
                            message: alloc::borrow::Cow::Borrowed("Hello World"),
                        },
                    ),
                    (
                        icu_provider::prelude::ResourceOptions {
                            variant: None,
                            langid: Some(icu_locid::LanguageIdentifier {
                                language: unsafe {
                                    icu_locid::subtags::Language::from_raw_unchecked(Some(28517))
                                },
                                script: None,
                                region: None,
                                variants: unsafe {
                                    icu_locid::subtags::Variants::from_raw_unchecked(None)
                                },
                            }),
                        },
                        icu_provider::hello_world::HelloWorldV1 {
                            message: alloc::borrow::Cow::Borrowed("Saluton, Mondo"),
                        },
                    ),
                    (
                        icu_provider::prelude::ResourceOptions {
                            variant: None,
                            langid: Some(icu_locid::LanguageIdentifier {
                                language: unsafe {
                                    icu_locid::subtags::Language::from_raw_unchecked(Some(24934))
                                },
                                script: None,
                                region: None,
                                variants: unsafe {
                                    icu_locid::subtags::Variants::from_raw_unchecked(None)
                                },
                            }),
                        },
                        icu_provider::hello_world::HelloWorldV1 {
                            message: alloc::borrow::Cow::Borrowed("سلام دنیا\u{200e}"),
                        },
                    ),
                    (
                        icu_provider::prelude::ResourceOptions {
                            variant: None,
                            langid: Some(icu_locid::LanguageIdentifier {
                                language: unsafe {
                                    icu_locid::subtags::Language::from_raw_unchecked(Some(26982))
                                },
                                script: None,
                                region: None,
                                variants: unsafe {
                                    icu_locid::subtags::Variants::from_raw_unchecked(None)
                                },
                            }),
                        },
                        icu_provider::hello_world::HelloWorldV1 {
                            message: alloc::borrow::Cow::Borrowed("hei maailma"),
                        },
                    ),
                    (
                        icu_provider::prelude::ResourceOptions {
                            variant: None,
                            langid: Some(icu_locid::LanguageIdentifier {
                                language: unsafe {
                                    icu_locid::subtags::Language::from_raw_unchecked(Some(29545))
                                },
                                script: None,
                                region: None,
                                variants: unsafe {
                                    icu_locid::subtags::Variants::from_raw_unchecked(None)
                                },
                            }),
                        },
                        icu_provider::hello_world::HelloWorldV1 {
                            message: alloc::borrow::Cow::Borrowed("Halló, heimur"),
                        },
                    ),
                    (
                        icu_provider::prelude::ResourceOptions {
                            variant: None,
                            langid: Some(icu_locid::LanguageIdentifier {
                                language: unsafe {
                                    icu_locid::subtags::Language::from_raw_unchecked(Some(24938))
                                },
                                script: None,
                                region: None,
                                variants: unsafe {
                                    icu_locid::subtags::Variants::from_raw_unchecked(None)
                                },
                            }),
                        },
                        icu_provider::hello_world::HelloWorldV1 {
                            message: alloc::borrow::Cow::Borrowed("こんにちは世界"),
                        },
                    ),
                    (
                        icu_provider::prelude::ResourceOptions {
                            variant: None,
                            langid: Some(icu_locid::LanguageIdentifier {
                                language: unsafe {
                                    icu_locid::subtags::Language::from_raw_unchecked(Some(24940))
                                },
                                script: None,
                                region: None,
                                variants: unsafe {
                                    icu_locid::subtags::Variants::from_raw_unchecked(None)
                                },
                            }),
                        },
                        icu_provider::hello_world::HelloWorldV1 {
                            message: alloc::borrow::Cow::Borrowed("Ave, munde"),
                        },
                    ),
                    (
                        icu_provider::prelude::ResourceOptions {
                            variant: None,
                            langid: Some(icu_locid::LanguageIdentifier {
                                language: unsafe {
                                    icu_locid::subtags::Language::from_raw_unchecked(Some(28530))
                                },
                                script: None,
                                region: None,
                                variants: unsafe {
                                    icu_locid::subtags::Variants::from_raw_unchecked(None)
                                },
                            }),
                        },
                        icu_provider::hello_world::HelloWorldV1 {
                            message: alloc::borrow::Cow::Borrowed("Salut,lume!"),
                        },
                    ),
                    (
                        icu_provider::prelude::ResourceOptions {
                            variant: None,
                            langid: Some(icu_locid::LanguageIdentifier {
                                language: unsafe {
                                    icu_locid::subtags::Language::from_raw_unchecked(Some(30066))
                                },
                                script: None,
                                region: None,
                                variants: unsafe {
                                    icu_locid::subtags::Variants::from_raw_unchecked(None)
                                },
                            }),
                        },
                        icu_provider::hello_world::HelloWorldV1 {
                            message: alloc::borrow::Cow::Borrowed("Привет, мир"),
                        },
                    ),
                    (
                        icu_provider::prelude::ResourceOptions {
                            variant: None,
                            langid: Some(icu_locid::LanguageIdentifier {
                                language: unsafe {
                                    icu_locid::subtags::Language::from_raw_unchecked(Some(26998))
                                },
                                script: None,
                                region: None,
                                variants: unsafe {
                                    icu_locid::subtags::Variants::from_raw_unchecked(None)
                                },
                            }),
                        },
                        icu_provider::hello_world::HelloWorldV1 {
                            message: alloc::borrow::Cow::Borrowed("Xin chào thế giới"),
                        },
                    ),
                    (
                        icu_provider::prelude::ResourceOptions {
                            variant: None,
                            langid: Some(icu_locid::LanguageIdentifier {
                                language: unsafe {
                                    icu_locid::subtags::Language::from_raw_unchecked(Some(26746))
                                },
                                script: None,
                                region: None,
                                variants: unsafe {
                                    icu_locid::subtags::Variants::from_raw_unchecked(None)
                                },
                            }),
                        },
                        icu_provider::hello_world::HelloWorldV1 {
                            message: alloc::borrow::Cow::Borrowed("你好世界"),
                        },
                    ),
                ],
            }
        };
        let data = &DATA_PROVIDER
            .load_payload(&DataRequest {
                resource_path: ResourcePath {
                    key: icu_provider::hello_world::key::HELLO_WORLD_V1,
                    options: ResourceOptions {
                        variant: None,
                        langid: Some(icu_locid::LanguageIdentifier {
                            language: unsafe {
                                icu_locid::subtags::Language::from_raw_unchecked(Some(25956))
                            },
                            script: None,
                            region: None,
                            variants: unsafe {
                                icu_locid::subtags::Variants::from_raw_unchecked(None)
                            },
                        }),
                    },
                },
            })
            .unwrap()
            .take_payload()
            .unwrap();
        {
            match (&data.get().message, &"Hallo Welt") {
                (left_val, right_val) => {
                    if !(*left_val == *right_val) {
                        let kind = ::core::panicking::AssertKind::Eq;
                        ::core::panicking::assert_failed(
                            kind,
                            &*left_val,
                            &*right_val,
                            ::core::option::Option::None,
                        );
                    }
                }
            }
        };
    }
    // and the same again
}
```

File: experimental/bakeable/Cargo.toml
Comment (Manishearth) [inline experimental/bakeable/Cargo.toml:6]: nit: probably should be crabbake-derive and in crabbake/derive, see how yoke does it

File: experimental/const_provider/src/lib.rs
Comment (Manishearth) [inline experimental/const_provider/src/lib.rs:105]: issue: this can just be `.clone()`, it's a shallow pointer copy for zero copy stuff

File: provider/blob/src/blob_data_provider.rs
Comment (Manishearth) [inline provider/blob/src/blob_data_provider.rs:140]: we should just

```
let zm = self.data.get();
zm.iter_keys()............. make a vec
```

File: experimental/bakeable/derive/src/lib.rs
Comment (Manishearth) [inline experimental/bakeable/derive/src/lib.rs:86]: useful to reexport this from your crate so that users don't have to import this

Comment (Manishearth) [reply experimental/bakeable/derive/src/lib.rs:86]:
```rust
#[doc(hidden)]
pub use proc_macro2::TokenStream
```

File: experimental/bakeable/derive/src/lib.rs
Comment (Manishearth) [reply experimental/bakeable/derive/src/lib.rs:48]: I'd also use `.add_bounds()` with `AddBounds::Fields`

https://docs.rs/synstructure/0.12.6/synstructure/struct.Structure.html#method.add_bounds

File: experimental/bakeable/src/lib.rs
Comment (Manishearth) [inline experimental/bakeable/src/lib.rs:27]: question: should there be such an impl for `str`? You can't actually generate a `str` literal

File: experimental/bakeable/src/lib.rs
Comment (Manishearth) [inline experimental/bakeable/src/lib.rs:29]: question: ditto, perhaps this should be on `&[T]`?

File: experimental/bakeable/src/lib.rs
Comment (Manishearth) [inline experimental/bakeable/src/lib.rs:82]: suggestion: i think this can be generic using `impl<T> Bakeable for Cow<'_, T> where &T: Bakeable`

**Review by Manishearth — CHANGES_REQUESTED** [review]
Mostly looks good so far! I haven't had a close look at the provider code, @sffc ?

I'm wondering if we should hold off on introducing any dependencies on bakeable to the rest of the code until the 0.5 release. We can still land this and experiment under experimental/

File: experimental/bakeable/src/lib.rs
Comment (robertbastian) [reply experimental/bakeable/src/lib.rs]: Can't figure out the lifetime bounds for this

File: experimental/bakeable/src/lib.rs
Comment (Manishearth) [inline experimental/bakeable/src/lib.rs]: Both of these need to be always exported since it's possible to use `bakeable_derive` as a direct crate dependency.

File: experimental/bakeable/src/lib.rs
Comment (Manishearth) [reply experimental/bakeable/src/lib.rs]: I think it needs to be `impl<'a, T> Bakeable for Cow<'a, T> where &'static T: Bakeable, T: 'static`

paired with `impl<T: Bakeable> Bakeable for &'static T`

Comment (robertbastian) [reply experimental/bakeable/src/lib.rs]: Nope. Could you give it a try?

Comment (Manishearth) [reply experimental/bakeable/src/lib.rs]: Done, pushed

Comment (robertbastian) [reply experimental/bakeable/src/lib.rs]: ah yes, `&&**` how could I miss that :D

File: experimental/bakeable/src/lib.rs
Comment (robertbastian) [reply experimental/bakeable/src/lib.rs]: Can we make it so it's not?

Comment (Manishearth) [reply experimental/bakeable/src/lib.rs]: No, we can at best suggest that people do not do this

But the additional `quote` dependency is not a huge deal IMO.

**Comment by iainireland** [comment]
Once this has matured a bit, I would be interested in an assessment of what requirements this places on provided data types, and a rough estimate of how well we are already meeting those requirements.

**Comment by Manishearth** [comment]
@iainireland it basically mandates that all data be zero-copy-compatible: no allocations, `Cow` and `zerovec` types for anything that would be otherwise allocated.

This is something we want to do anyway, but we haven't been working on that as much.

We can design this so that it is possible for data to temporarily not be crabbake-compatible (and thus not usable from a crabbake data provider) to aid velocity where needed.

**Comment by iainireland** [comment]
Right, what I'm particularly interested in is any cases where it's not immediately obvious that a type isn't zero-copy compatible, or not completely trivial to switch over.

Basically my angle on this is that, if ICU4X is going to succeed in the long run, it's going to need to be able to onboard new developers, after many people currently working on it have moved on. Stuff like Yoke and ZeroVec solve real problems, but they also add to the complexity budget of things that people need to figure out at some point as they learn the codebase. The delta between crabbake-compatible types and the types we wrote naturally without crabbake in mind seems like a reasonable proxy for developer brain overhead.

I'm optimistic that crabbake will be smooth sailing, but I want to make sure that we're at least thinking about it a little.

**Comment by Manishearth** [comment]
> The delta between crabbake-compatible types and the types we wrote naturally without crabbake in mind seems like a reasonable proxy for developer brain overhead.


So to be clear anything that is not crabbake-compatible currently uses `#[yoke(cloning_zcf)]`. ZeroCopyFrom is no longer necessary and we'll eventually get rid of those derives, but for now the `cloning_zcf` serves as a marker, if you want to go check what has a delta in crabbake. The trickiest things will be when someone wishes to use an enum inside a ZeroVec/etc since those cases are complex. Otherwise, as long as the ZeroVec proc macro exists (not yet, but soon), we should be good.

**Comment by Manishearth** [comment]
I suspect the path forward might be to land this as experimental, flesh out impls for zerovec/etc, and use this as an impetus to drive the conversion of the rest of the project to zero-copy.

**Comment by Manishearth** [comment]
@sffc I think we're waiting on a data provider review from you to land this, fwiw. The PR looks good from my end.

**Comment by sffc** [comment]
I want to land the changes proposed in #1479 first, because they will lay the groundwork for the CrabBake data provider implementation.

**Comment by robertbastian** [comment]
I've removed the data provider, this is now just the `Bakeable` trait and derive

**Review by Manishearth — APPROVED** [review]

**Review by sffc — APPROVED** [review]
Nice job!

The next thing we should do, even before adding more features, is to get a proper testing harness set up.  We should use one similar to what we use in Diplomat, which generates golden output files from CrabBake which helps with test coverage, ease of adding new tests, and ease of viewing the test expectations.

**Review by Manishearth — APPROVED** [review]

**Review by sffc — COMMENTED** [review]
post-submit LGTM

**Pull request merged**
